### PR TITLE
enable database custom alias configuration #4

### PR DESCRIPTION
**Pull request by SaschaJohn**
@bsideup @rnorth 
my approach so solve #4 
I also added documenttaion and testcases.
Looking forward for feedback.

Regards

Sascha

### CODE REVIEW

File: modules/jdbc-test/src/test/java/org/testcontainers/jdbc/JDBCDriverAliasPropertiesTest.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+
+import com.zaxxer.hikari.HikariConfig;
+import com.zaxxer.hikari.HikariDataSource;
+
+@RunWith(Parameterized.class)
+public class JDBCDriverAliasPropertiesTest {
```
Comment (bsideup) [inline modules/jdbc-test/src/test/java/org/testcontainers/jdbc/JDBCDriverAliasPropertiesTest.java]: This class should be a unit test. We should test that aliases map to the right db type/image, there is no need to actually start the DB

File: modules/jdbc-test/src/test/java/org/testcontainers/jdbc/JDBCDriverAliasPropertiesTest.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                {"jdbc:tc:mysqlserver://hostname:hostport;databaseName=databasename", EnumSet.noneOf(Options.class)}                
+            });
+    }
+    
+    @Test
+    public void shouldInstanciateContainersAccordingToPropertiesWithDefaultImageTag() throws SQLException {
+    	Properties p = new Properties();
+    	p.put("jdbc.alias.mysqlserver.type", "sqlserver");
+    	p.put("jdbc.alias.mysqlserver.image", "mcr.microsoft.com/mssql/server");
+    	ContainerDatabaseDriver.clearJBDCAliasProperties();
```
Comment (bsideup) [inline modules/jdbc-test/src/test/java/org/testcontainers/jdbc/JDBCDriverAliasPropertiesTest.java]: this should use the configuration mock, see other configuration-sensitive tests (e.g. tests for the reusable mode)

File: modules/jdbc-test/src/test/java/org/testcontainers/jdbc/JDBCDriverAliasTest.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+
+import com.zaxxer.hikari.HikariConfig;
+import com.zaxxer.hikari.HikariDataSource;
+
+@RunWith(Parameterized.class)
+public class JDBCDriverAliasTest {
```
Comment (bsideup) [inline modules/jdbc-test/src/test/java/org/testcontainers/jdbc/JDBCDriverAliasTest.java]: should also be a unit test

File: modules/jdbc-test/src/test/java/org/testcontainers/jdbc/JDBCDriverTest.java [hunk 4]
```diff
@@ -110,7 +110,7 @@ public void test() throws SQLException {
         }
     }
 
-    private void performSimpleTest(HikariDataSource dataSource) throws SQLException {
+    protected void performSimpleTest(HikariDataSource dataSource) throws SQLException {
```
Comment (bsideup) [inline modules/jdbc-test/src/test/java/org/testcontainers/jdbc/JDBCDriverTest.java]: unrelated?

File: modules/jdbc-test/src/test/resources/container-license-acceptance.txt [hunk 5]
```diff
@@ -1,2 +1,3 @@
 mcr.microsoft.com/mssql/server:2017-CU12
+mcmoe/mssqldocker:latest
```
Comment (bsideup) [inline modules/jdbc-test/src/test/resources/container-license-acceptance.txt]: please remove after converting into a unit test, we should not accept licenses of 3rd party images

File: modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java [hunk 6]
```diff
@@ -40,12 +41,23 @@
     private Driver delegate;
     private static final Map<String, Set<Connection>> containerConnections = new HashMap<>();
     private static final Map<String, JdbcDatabaseContainer> jdbcUrlContainerCache = new HashMap<>();
+    private static final Map<String, Map<String, String>> jdbcAliasMap = new HashMap<String, Map<String,String>>();
```
Comment (bsideup) [inline modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java]: please make Map's value a type, not just `Map<String, String>`

File: modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java [hunk 7]
```diff
@@ -40,12 +41,23 @@
...
     static {
         load();
     }
+    
+    /**
+     * Parse properties in default constructor
+     */
+    public ContainerDatabaseDriver() {
+    	super();
+    	mapJBDCAliasProperties(TestcontainersConfiguration.getInstance().getProperties());
```
Comment (bsideup) [inline modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java]: Why cached? You can get the property when you encounter an alias

File: modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java [hunk 8]
```diff
@@ -73,8 +85,8 @@ public synchronized Connection connect(String url, final Properties info) throws
         ConnectionUrl connectionUrl = ConnectionUrl.newInstance(url);
 
         synchronized (jdbcUrlContainerCache) {
-
-            String queryString = connectionUrl.getQueryString().orElse("");
+        	
+        	String queryString = connectionUrl.getQueryString().orElse("");
```
Comment (bsideup) [inline modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java]: please fix the formatting

File: modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java [hunk 9]
```diff
@@ -227,6 +255,52 @@ private void runInitFunctionIfRequired(final ConnectionUrl connectionUrl, Connec
...
         }
     }
+    
+    /**
+     * Returns the tag for a JBDC alias if set or the default one from the implementing class 
+     * @return tag
+     */
+    private String getAliasTag(Map<String, String> aliasMap, JdbcDatabaseContainer container) {
+    	String tag = aliasMap.get("tag");
+		if (tag == null) { // no tag was set, so use the default one
```
Comment (bsideup) [inline modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java]: wrong formatting

File: modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java [hunk 10]
```diff
@@ -91,9 +103,25 @@ public synchronized Connection connect(String url, final Properties info) throws
...
+                		type = aliasMap.get("type"); //overwrite type from url with type from alias definition
+                		wasAlias = true;
+                	}
+                	
+                    if (candidateContainerType.supports(type)) {
+                    	
+                    	container = candidateContainerType.newInstance(connectionUrl);
+                    	
+                    	if (wasAlias) { //overwrite image and tag
+                    		container.setDockerImageName(aliasMap.get("image")+":"+getAliasTag(aliasMap, container));
```
Comment (bsideup) [inline modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java]: This does not feel like the right place to do it. Consider handling aliases in `JdbcDatabaseContainerProvider#newInstance`. We may also need to introduce a new API there to handle new instance creation with custom image

File: modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java [hunk 11]
```diff
@@ -298,4 +372,31 @@ static JdbcDatabaseContainer getContainer(String jdbcUrl) {
...
+    	ContainerDatabaseDriver.jdbcAliasMap.get(alias).put("tag", tag);			
+	}
+    
+    /**
+     * Register JDBC alias for custom JDBC-URL 
+     * @param alias		name of the alias e.g "my-own-mysql"
+     * @param type		driver type of the alias specified e.g. "mysql", "sqlserver", ...
+     * @param image		string that represents the docker image e.g. "mysql", "customregistry/image name"
+     */
+    public static void registerAlias(String alias, String type, String image) {
```
Comment (bsideup) [inline modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java]: since we discourage the use of `:latest`, we should not provide any tag-less API

Comment (SaschaJohn) [reply modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java]: @bsideup 
this will not fallback to latest but to the tag, that was specified by the module maintainer in the underlying specific container class.
The usecase here is when you host your own proxy registry, but want to use the tag, that was specified already.  Somewhat related to feture request in #2100 at least for the db containers

File: modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java (hunk 10 above)
Comment (SaschaJohn) [reply modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java]: @bsideup 
newInstance() is overridden in implementing modules, I didn't want to touch existing modules.
For me it's the smartest way to overwrite the imageName right after instance creation.

File: modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java (hunk 7 above)
Comment (SaschaJohn) [reply modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java]: @bsideup we don't know the aliases without reading the configuration at instanciation.
IMHO it's the easiest way to implement it here, otherwise we won't be able to find candiate containers with the alias and have to see if the name is an alias afterwards...

Comment (bsideup) [reply modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java]: but you don't need it at instantiation time.

The algorithm should be something like:
1. get the type from the JDBC URL
2. try to get alias by it
3. if not found, fallback to the default one (the behaviour before this PR)

File: modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java (hunk 10 above)
Comment (bsideup) [reply modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java]: this is why I was talking about introducing a new API. `setDockerImageName` here is a bit hacky and we should expose first-class support for creating JDBC containers from a custom image

Comment (SaschaJohn) [reply modules/jdbc/src/main/java/org/testcontainers/jdbc/ContainerDatabaseDriver.java]: @bsideup the Container is created according to the property DockerImageName. The property as a public setter isn't that an API ;-)

**Comment by SaschaJohn** [comment]
@binkley, without specifying image tag the tag from the original implementing class is used,
for your example default the version, from the Postgres Container class.
With image tag, the code will use your explicit tag configuration. Please see also documentation part in the PR

**Comment by rnorth** [comment]
@bsideup I know you were grappling with a more difficult version of this problem for r2dbc support. Would you like to take over / replace this PR?
